File: src/Controls/MIDIKeyboard.ts
import type { getHandleMidi } from '../Instruments/Synth';
import { Commands } from '../Tools/Midi';
import type { adjustContinuousControl } from './ContinuousControl';

export async function listen(
  handleMidi: ReturnType<typeof getHandleMidi>,
  adjustContinuousControlFunction: typeof adjustContinuousControl,
) {
  const midiAccess: WebMidi.MIDIAccess = await navigator.requestMIDIAccess();

  const destructors = [...midiAccess.inputs.entries()].reduce<
    Array<() => void>
  >((carry, [, input]) => {
    // Remove old listener if it exists
    const listener = (e: WebMidi.MIDIMessageEvent) => {
      const [command, message, value] = e.data;

      switch (command) {
        case Commands.NOTE_ON:
        case Commands.NOTE_OFF:
          handleMidi({ command, message, value });
          break;
        case Commands.CONTINUOUS_CONTROL:
          adjustContinuousControlFunction(message, value);
          break;
        default:
          break;
      }
    };

    input.addEventListener('midimessage', listener);

    function destruct() {
      input.removeEventListener(
        'midimessage',
        listener as unknown as EventListener,
      );
    }

    return [...carry, destruct];
  }, []);

  return () => {
    destructors.forEach(destruct => destruct());
  };
}

File: web/Controls/MidiKeyboard.dart
import 'dart:html';

import '../Instruments/Synth.dart';
import '../Tools/Midi.dart';

void listen({required IPlayTone playTone}) async {
  final midiAccess = await window.navigator.requestMidiAccess() as MidiAccess;

  var stopTones = <num, void Function()>{};

  midiAccess.inputs?.forEach((String name, dynamic entry) {
    final midiInput = entry as MidiInput;

    midiInput.onMidiMessage.listen((event) {
      if (event.data == null) return;

      final command = event.data![0];
      final message = event.data![1];
      final value = event.data![2];

      switch (command) {
        case Commands.NOTE_ON:
          stopTones[message] = playTone(message: message, velocity: value);
          break;
        case Commands.NOTE_OFF:
          if (stopTones[message] == null) return;
          stopTones[message]!();
          break;
        default:
          break;
      }
    });
  });
}
